# PBBtools: bioinformatic tools developed by the Plant Biotechnology and Bioinformatics group @ TU Braunschweig
This is a collection of tools developed by the [Plant Biotechnology and Bioinformatics](https://www.tu-braunschweig.de/en/ifp/pbb) group @ TU braunschweig. Many of these tools are available via our [BioinfToolServer](https://www.tu-braunschweig.de/en/ifp/pbb/tools).

# KIPEs: Knowledge-based Identification of Pathway Enzymes
**Description**: This tool facilitates the identification of genes/proteins involved in a particular pathway. KIPEs was initially developed for the annotation of the flavonoid biosynthesis, but can also be applied to other pathways.

**Link**: [KIPEs repository](https://github.com/bpucker/KIPEs) and [KIPEs webserver](http://pbb.bot.nat.tu-bs.de/KIPEs/)

**Reference**: Rempel A & Pucker B (2022). KIPEs3: Automatic annotation of biosynthesis pathways. bioRxiv 2022.06.30.498365; doi:[10.1101/2022.06.30.498365](https://doi.org/10.1101/2022.06.30.498365).


# MYB_annotator: Automatic identification and annotation of MYB transcription factors


# CoExp: Co-expression analysis


# Identification of Open Reading Frame
This script screens given nucleotide sequences for the longes Open Reading Frame (ORF) / coding sequence (CDS) that could encode a peptide sequence. The longest detected ORF/CDS per input sequence is reported.

https://github.com/bpucker/PBBTools/blob/main/cds_finder/cds_finder_readme.md


# Translation of CDS into peptide sequence
This script reads all sequences from a given FASTA file and converts them into the corresponding peptide sequences.

https://github.com/bpucker/PBBTools/blob/main/transeq/transeq_readme.md


# References
Rempel A & Pucker B (2022). KIPEs3: Automatic annotation of biosynthesis pathways. bioRxiv 2022.06.30.498365; doi:[10.1101/2022.06.30.498365](https://doi.org/10.1101/2022.06.30.498365).
