# PBBtools: bioinformatic tools developed by the Plant Biotechnology and Bioinformatics group @ TU Braunschweig
This is a collection of tools developed by the [Plant Biotechnology and Bioinformatics](https://www.tu-braunschweig.de/en/ifp/pbb) group @ TU braunschweig. Many of these tools are available via our [BioinfToolServer](https://www.tu-braunschweig.de/en/ifp/pbb/tools).



# KIPEs: Knowledge-based Identification of Pathway Enzymes
**Description**: This tool facilitates the identification of genes/proteins involved in a particular pathway. KIPEs was initially developed for the annotation of the flavonoid biosynthesis, but can also be applied to other pathways.

**Link**: [KIPEs repository](https://github.com/bpucker/KIPEs) and [KIPEs webserver](http://pbb.bot.nat.tu-bs.de/KIPEs/)

**Reference**: Rempel A & Pucker B (2022). KIPEs3: Automatic annotation of biosynthesis pathways. bioRxiv 2022.06.30.498365; doi:[10.1101/2022.06.30.498365](https://doi.org/10.1101/2022.06.30.498365).



# MYB_annotator: Automatic identification and annotation of MYB transcription factors
**Description**: This tool facilitates the identification of MYB transcription factors in a given species. Identified MYBs are assigned to well characterized orthologs and a functional annotation is transferred.

**Link**: [MYB_annotator repository](https://github.com/bpucker/MYB_annotator) and [MYB_annotator webserver](http://pbb.bot.nat.tu-bs.de/MYB_annotator/)

**Reference**: Pucker, B. Automatic identification and annotation of MYB gene family members in plants. BMC Genomics 23, 220 (2022). doi:[10.1186/s12864-022-08452-5](https://doi.org/10.1186/s12864-022-08452-5).



# CoExp: Co-expression analysis
**Description**: This tool performs a pairwise co-expression analysis for a given set of genes of interest. All co-expressed genes are identified for this set.

**Link**: [CoExp repository](https://github.com/bpucker/CoExp) and [CoExp webserver](http://pbb.bot.nat.tu-bs.de/CoExp/)

**Reference**: This repository.



# Identification of Open Reading Frame
**Description**: This script screens given nucleotide sequences for the longes Open Reading Frame (ORF) / coding sequence (CDS) that could encode a peptide sequence. The longest detected ORF/CDS per input sequence is reported.

**Link**: [CDS_finder repository](https://github.com/bpucker/PBBTools/blob/main/cds_finder/cds_finder_readme.md)

**Reference**: This repository.



# Translation of CDS into peptide sequence
**Description**: This script reads all sequences from a given FASTA file and converts them into the corresponding peptide sequences.

**Link**: [transeq repository](https://github.com/bpucker/PBBTools/blob/main/transeq/transeq_readme.md)

**Reference**: This repository.



# MGSE: Mapping-based Genome Size Estimation
**Description**: This script estimates the size of a plant genome based on the coverage in a read mapping.

**Link**: [MGSE repository](https://github.com/bpucker/MGSE)

**Reference**: Pucker B. Mapping-based genome size estimation. bioRxiv 607390; doi:[10.1101/607390](https://doi.org/10.1101/607390).



# NAVIP: Neighborhood-Aware Variant Impact Predictor
**Description**: This script predicts the functional impact of sequence variants on the surrounding gene. NAVIP considers all variants in a gene simultaneously to account for their interactions.

**Link**: [NAVIP repository](https://github.com/bpucker/NAVIP)

**Reference**: Baasner, J.-S., Howard, D., Pucker, B.(2019). Influence of neighboring small sequence variants on functional impact prediction. bioRxiv. doi:[10.1101/596718](https://doi.org/10.1101/596718).



# Long Read Walker



# References
Rempel A & Pucker B (2022). KIPEs3: Automatic annotation of biosynthesis pathways. bioRxiv 2022.06.30.498365; doi:[10.1101/2022.06.30.498365](https://doi.org/10.1101/2022.06.30.498365).
